src/mpid/ch4/src/ch4r_win.h

```diff
@@ -515,7 +500,8 @@ static inline int MPIDI_CH4R_mpi_win_lock(int lock_type, int rank, int assert, M
 
     MPIDI_CH4U_LOCK_EPOCH_CHECK_NONE(win, rank, mpi_errno, goto fn_fail);
 
-    MPIDI_CH4U_win_target_t *target_ptr = &MPIDI_CH4U_WIN(win, targets)[rank];
+    MPIDI_CH4U_win_target_t *target_ptr = MPIDI_CH4U_win_target_add(win, rank);
+
```

Comment:
Thanks for making this optimization. Let me try to summarize the design of target object allocation. Please correct me if my understanding is wrong:
1. Always create a target at `lock`, and free it at `unlock`.
2. For `lock_all`, `lock`, or `start-complete` epoch, create a target object at operation issuing time if such object does not exist.
3. Always clean all target objects in window at `unlock_all`, `win_complete`.
4. In `fence`, clean all target objects in window if `assert` does not include `MPI_MODE_NOPRECEDE`.

Here are my questions:
1. This design is very similar to the memory optimizing policy in CH3. But I am concerning performance drop if we exactly follow the current policy in CH4 which is much more latency sensitive than CH3. To be specific, for a program that does frequent lock-unlock on same group of targets, creation-freeing target at every epoch would obviously increase latency. Can we reuse the allocated target object, and just release all at `win_free` ? Or, at least we should allow user to pass `info hint` to us whether the program prefers a `memory efficient` mode or a `performance efficient` mode.
2. Why clean all target objects in `fence without MPI_MODE_NOPRECEDE` ? Any existing target object should be freed by the previous epoch closing call (if the user program is correct), right ?

Reply:
@min-si , your understanding is correct.
For your questions,
1. It is a trade-off between memory cost and lock/unlock cost.
I talked with @hajimefu , yes, one option is reusing existing target obj and freeing all at `win_free`. But in this case, an application requiring locking for multiple targets (just an extreme opposite example to your case) would introduce high memory cost.

My suggestion is using current implementation, and create an enhancement ticket about `info hint` (I can fix that when my current workload is done, should not take much time).

Anyway, it is up to you.

2. For `win_fence`, my understanding is that `MPI_MODE_NOPRECEDE` assertion means `no RMA completion` (from book `Using advanced MPI`). So there is no need to clean all existing `targets` as all ops are supposed not to be done.

Reply:
1. The `info hint` plan sounds good to me. Thanks !

2. In `fence` epoch, your implementation does not add per-target object at operation issuing time,  so you should not need the cleanup at a `fence` call. The only case that would cause existence of per-target objects at `fence` epoch is incorrect user program, e.g., user calls `start`->'put (per-target object is created here)'->'fence'. But the `MPIDI_CH4U_FENCE_EPOCH_CHECK` sync check should catch such case, and throw error if user does not disable `error-checking` at MPICH configure time. If user disables `error-checking`, an incorrect user program causes undefined behavior, thus you do not need to free those target objects either.